app/src/ingest_bem_pdfs.py

```diff
@@ -76,15 +78,71 @@ def _parse_pdf(file: BinaryIO) -> list[EnrichedText]:
     for text in grouped_texts:
         text.id = str(uuid.uuid1())
     assert len(set(text.id for text in grouped_texts)) == len(grouped_texts)
+
     return grouped_texts
 
 
-def enrich_texts(file: BinaryIO) -> list[EnrichedText]:
-    "Placeholder function. Will be implemented for DST-414, probably in a different file."
+def _enrich_texts(file: BinaryIO) -> list[EnrichedText]:
+    unstuctured_elem_list = partition_pdf(file=file, strategy="fast")
+    enrich_text_list = []
+
     outline: list[Heading] = pdf_utils.extract_outline(file)
-    for _heading in outline:
-        pass
-    return []
+    current_headings: list[Heading] = []
+    for element in unstuctured_elem_list:
+        if element.category == "Footer" or element.category == "Header":
+            continue
+
+        # Unstructured fails to categorize the date strings in the header,
+        # so manually check for that and ignore those too
+        if element.category == "UncategorizedText" and re.match(
+            r"^\d{1,2}-\d{1,2}-\d{4}$", element.text
+        ):
+            continue
+
+        if element.category == "Title":
+            current_headings = _get_current_heading(outline, element, current_headings)
+        try:
+            enriched_text_item = EnrichedText(
+                text=element.text,
+                type=TextType(element.category),
+                page_number=element.metadata.page_number,
+                headings=current_headings,
+                id=element._element_id,
+            )
+            enrich_text_list.append(enriched_text_item)
+        except ValueError:
+            logger.warning(
+                f"{element.category} is not an accepted TextType, {element.text}, {element.metadata.page_number}"
+            )
+    return enrich_text_list
+
+
+def _match_heading(
+    outline: list[Heading], heading_name: str, page_number: int | None
+) -> Heading | None:
+    for heading in outline:
+        if heading.pageno == page_number:
+            # account for spacing differences in unstructured and pdfminer parsing
+            heading_words = [word for word in heading.title.casefold() if not word.isspace()]
+            element_words = [word for word in heading_name.casefold() if not word.isspace()]
+            if heading_words == element_words:
+                return heading
+    return None
+
+
+def _get_current_heading(
+    outline: list[Heading], element: Element, current_headings: list[Heading]
+) -> list[Heading]:
+    if heading := _match_heading(outline, element.text, element.metadata.page_number):
+        if heading.level == 1:
+            current_headings = [heading]
+        else:
+            if heading.title != current_headings[-1].title:
+                current_headings = current_headings[: -(len(current_headings) - 1 - heading.level)]
```

Comment:
Why isn't this just:
```suggestion
                current_headings = current_headings[:heading.level]
```
i.e., copy all the headings up to the new `heading`.

Reply:
there's an off by one issue if it's not negative, revised in latest commit